src/response_file.ml

```diff
@@ -0,0 +1,54 @@
+open Import
+
+(* After seeing this number of paths, switch to a response file *)
+let max_paths = 4
+
+let too_long spec =
+  let add n len =
+    let n = n + len in
+    if n >= max_paths then raise_notrace Exit;
+    n
+  in
+  let rec loop acc (spec : Arg_spec.Simple.t) =
+    match spec with
+    | A _ | As _ -> acc
+    | Path _ | Dep _ -> add acc 1
+    | Paths l | Deps l -> add acc (List.length l)
+    | S l -> List.fold_left l ~init:acc ~f:loop
+  in
+  try
+    ignore (loop 0 spec : int);
+    false
+  with Exit ->
+    true
+
+let process sctx ~key ~exec_dir ~response_file_dir ~arg_name
+      ~dump_response_file spec =
+  match arg_name with
+  | None -> Arg_spec.of_simple spec
+  | Some arg ->
+    if not (Sys.win32 && too_long spec) then
```

Comment:
Rather than hardcoding `Sys.win32` & a number of paths, can we have a function that guesses the maximum number depending on the platform? (e.g. `if Sys.win32 then Some 8000 else None`)? Also, there might be some cross compiling concerns here - I'm not sure whether the target toolchain has to be queried as well.

Reply:
Cross-compilation shouldn't be an issue. It's really about the processes that are executed on the current machine.